tools/setup.py

```diff
@@ -44,29 +45,68 @@ def write_lastchange():
     # ])
 
 
-def get_gn_args():
+# If this text is found in args.gn, we assume it hasn't been hand edited.
+gn_args_header = [
+    "# This file is automatically generated by tools/build.py.",
+    "# REMOVE THIS LINE to preserve any changes you make.", ""
+]
+
+
+def gn_args_are_generated(lines):
+    for line in lines:
+        if re.match("^\s*#.*REMOVE THIS LINE", line):
+            return True
+    return False
+
+
+def read_gn_args(args_filename):
+    if not os.path.exists(args_filename):
+        return (None, False)  # No content, not hand edited.
+
+    with open(args_filename) as f:
+        lines = f.read().splitlines()
+        args = [l.strip() for l in lines if not re.match("^\s*(#|$)", l)]
+        hand_edited = not gn_args_are_generated(lines)
+        return (args, hand_edited)
+
+
+def write_gn_args(args_filename, args):
+    assert not gn_args_are_generated(args)  # No header -> hand crafted.
+    lines = gn_args_header + args
+    assert gn_args_are_generated(lines)  # With header -> generated.
+
+    # Ensure the directory where args.gn goes exists.
+    dir = os.path.dirname(args_filename)
+    if not os.path.isdir(dir):
+        os.makedirs(dir)
+
+    with open(args_filename, "w") as f:
+        f.writelines("\n".join(lines) + "\n")
+
+
+def generate_gn_args(mode):
     out = []
-    if build_mode() == "release":
+    if mode == "release":
         out += ["is_official_build=true"]
-    elif build_mode() == "debug":
+    elif mode == "debug":
         pass
     else:
-        print "Bad mode {}. Use 'release' or 'debug' (default)" % build_mode()
+        print "Bad mode {}. Use 'release' or 'debug' (default)" % mode
         sys.exit(1)
+
     if "DENO_BUILD_ARGS" in os.environ:
         out += os.environ["DENO_BUILD_ARGS"].split()
 
     # Check if ccache or sccache are in the path, and if so we set cc_wrapper.
     cc_wrapper = find_executable("ccache") or find_executable("sccache")
     if cc_wrapper:
-        out += [r'cc_wrapper="%s"' % cc_wrapper]
-        # Windows needs a custom toolchain for cc_wrapper to work.
+        out += ['cc_wrapper="%s"' % cc_wrapper]
+        # For cc_wrapper to work on Windows, we need need a to select our own
+        # toolchain by overriding 'custom_toolchain' and 'host_toolchain'.
+        # TODO: Is there a way to use it without the involvement of args.gn?
```

Comment:
Any theories as to why it has to be done in args.gn ?

Reply:
Well I can't set default_toolchain elsewhere. 
Except in .gn in the root but the current_os, is_win etc variables aren't available there.
So where else would it go?